servicemodels/e2sm_rc/pdubuilder/E2SM-RcRanfunctionDefinition.go

```diff
@@ -0,0 +1,93 @@
+//  SPDX-FileCopyrightText: 2020-present Open Networking Foundation <info@opennetworking.org>
+//
+//  SPDX-License-Identifier: Apache-2.0
+
+package pdubuilder
+
+import (
+	e2smcommonies "github.com/onosproject/onos-e2-sm/servicemodels/e2sm_rc/v1/e2sm-common-ies"
+	e2smrcv1 "github.com/onosproject/onos-e2-sm/servicemodels/e2sm_rc/v1/e2sm-rc-ies"
+)
+
+func CreateE2SmRcRanfunctionDefinition(rfsn string, oID string, dscr string) (*e2smrcv1.E2SmRcRanfunctionDefinition, error) {
+
+	msg := &e2smrcv1.E2SmRcRanfunctionDefinition{
+		RanFunctionName: &e2smcommonies.RanfunctionName{
+			RanFunctionShortName:   rfsn,
+			RanFunctionE2SmOid:     oID,
+			RanFunctionDescription: dscr,
+		},
+	}
+
+	return msg, nil
+}
+
+func CreateE2SmRcRanfunctionDefinitionEventTrigger(rfsn string, oID string, dscr string, et *e2smrcv1.RanfunctionDefinitionEventTrigger) (*e2smrcv1.E2SmRcRanfunctionDefinition, error) {
+
+	msg := &e2smrcv1.E2SmRcRanfunctionDefinition{
+		RanFunctionName: &e2smcommonies.RanfunctionName{
+			RanFunctionShortName:   rfsn,
+			RanFunctionE2SmOid:     oID,
+			RanFunctionDescription: dscr,
+		},
+		RanFunctionDefinitionEventTrigger: et,
+	}
+
+	return msg, nil
+}
+
+func CreateE2SmRcRanfunctionDefinitionReport(rfsn string, oID string, dscr string, report *e2smrcv1.RanfunctionDefinitionReport) (*e2smrcv1.E2SmRcRanfunctionDefinition, error) {
+
+	msg := &e2smrcv1.E2SmRcRanfunctionDefinition{
+		RanFunctionName: &e2smcommonies.RanfunctionName{
+			RanFunctionShortName:   rfsn,
+			RanFunctionE2SmOid:     oID,
+			RanFunctionDescription: dscr,
+		},
+		RanFunctionDefinitionReport: report,
+	}
+
+	return msg, nil
+}
+
+func CreateE2SmRcRanfunctionDefinitionInsert(rfsn string, oID string, dscr string, insert *e2smrcv1.RanfunctionDefinitionInsert) (*e2smrcv1.E2SmRcRanfunctionDefinition, error) {
+
+	msg := &e2smrcv1.E2SmRcRanfunctionDefinition{
+		RanFunctionName: &e2smcommonies.RanfunctionName{
+			RanFunctionShortName:   rfsn,
+			RanFunctionE2SmOid:     oID,
+			RanFunctionDescription: dscr,
+		},
+		RanFunctionDefinitionInsert: insert,
+	}
+
+	return msg, nil
+}
+
+func CreateE2SmRcRanfunctionDefinitionControl(rfsn string, oID string, dscr string, control *e2smrcv1.RanfunctionDefinitionControl) (*e2smrcv1.E2SmRcRanfunctionDefinition, error) {
+
+	msg := &e2smrcv1.E2SmRcRanfunctionDefinition{
+		RanFunctionName: &e2smcommonies.RanfunctionName{
+			RanFunctionShortName:   rfsn,
+			RanFunctionE2SmOid:     oID,
+			RanFunctionDescription: dscr,
+		},
+		RanFunctionDefinitionControl: control,
+	}
+
+	return msg, nil
+}
+
+func CreateE2SmRcRanfunctionDefinitionPolicy(rfsn string, oID string, dscr string, policy *e2smrcv1.RanfunctionDefinitionPolicy) (*e2smrcv1.E2SmRcRanfunctionDefinition, error) {
```

Comment:
We should be able to set all Policy, Control, Insert definitions part of one ran function description. Why they are separate? Are you using builder patterns?

Reply:
Would be nice to have a unit test which shows how to build a complete ran function description which can be included in E2 setup and it is consistent with encoding of optional/non optional fields (i.e. the minimum that we should set that we can include it in E2 setup. )

Reply:
Look at this example in ran-simulator for kpmv2 https://github.com/onosproject/ran-simulator/blob/master/pkg/utils/e2sm/kpm2/ranfuncdescription/ranfuncdescription.go#L82

Reply:
Agree. All lists included in the RanfunctionDescription are optional, so I assume that their presence may vary depending on the use-case (i.e., event trigger, report, policy, etc.). 

Builder pattern for RFD to come in next push. For all other SMs to come in the next PR to unblock you.
Unit tests are also to come in the next PR.